content/2019-03-26_dvacaty_treti_sraz.md

```diff
@@ -0,0 +1,30 @@
+Modified: 2019-03-26 18:00:00
+Author: Anežka Müller
+
+V rámci retrospektivy, co se událo za poslední týden, jsme narazily na několik nejasností v rámci tříd. V našem projektu hrají třídy podstatnou roli, umožňují nám celý kód výrazně zjednodušit, zpřehlednit a zefektivnit. Proto jsme se rozhodly pro malé opakování, abychom si osvěžily informace ze [začátečnického kurzu](https://naucse.python.cz/2019/brno-jaro-2019-pondeli/beginners/class/).
+
+Při vytváření tříd je neopomenutelným parametrem jejích atributů a metod `self`, které vyjadřuje konkrétní objekt dané třídy. Při volání metody na konkrétní objekt pak metoda místo `self` dostává objekt, na kterém je metoda volána.
+
+Konkrétním objektům třídy můžeme definovat atributy i bez toho, aby byly "vynucené" metodou `__init__`, která říká, jaké atributy musí mít všechny objekty dané třídy a které je třeba nadefinovat již při vytváření objektu. 
+
+Stejně jako u jiných částí kódu, i u metod třídy je dobré se dívat, zda něco zbytečně neduplikujeme. Z hlediska fází hry nebo pojmenování nějaké metody může dávat smysl mít části rozdělené a metody pak volat podle názvu v té fázi hry, kde se lépe hodí, pokud však mají většinu obsahu shodnou, je dobré se zamyslet nad tím, zda by nebylo lépe kód sjednotit pod jednu metodu, která bude obecněji uplatnitelná. Příkladem z našeho kódu budiž metody `Walk` a `Move`, které obě patří třídě `Robot`. Jednu používáme, pokud se robot pohybuje na základě programu, který mu zadáme pomocí karet, druhou pak, pokud robotem hýbe nějaká jiná vnější příčina. Každá sice dělá něco jiného a používá se v jiné situaci, ale vnitřek se liší pouze tím, že v druhém případě se metoda neprovede, pokud na políčku, na které by se měl robot posunout, již nějaký robot stojí. (Dle pravidel při tomto pohybu nemá robot dostatečnou sílu na to pohnout jiným robotem a odsunout jej.) Pokud tedy ošetříme tuto část, můžeme nechat plný kód pouze v jedné z metod a v druhé ji pak pouze zavolat se správnými argumenty.
+Případně je ještě možné společnou část kódu vydělit do pomocné funkce (označujeme s `_` na začátku názvu funkce) a tu pak použít v obou původních funkcích.
+
+### OrderedDict
+
+Doposud jsme v kódu pracovaly se seznamem startovních políček bez jakýchkoliv dalších detailů. Pro hru ale potřebujeme krom polohy startovních políček ještě několik doplňujícíh informací, například kterým směrem jsou orientována, a také potřebujeme, aby záznamy zůstávaly ve stejném pořadí. Karolina se rozhodla předělat seznam na slovník, kde klíčem bude číslo políčka a hodnotami budou směr a souřadnice. Starší verze Pythonu ale náhodně mění pořadí položek ve slovníku, bylo tedy třeba použít knihovnu `Collections` a z ní třídu [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict), která starším verzím Pythonu přidávám schopnost držet záznamy ve slovníku v takovém pořadí, v jakém jsou do něj vkládány. Od Pythonu 3.7 výše je tato vlastnost slovníků přímo zaintegrovaná do Pythonu. 
```

Comment:
```suggestion
Doposud jsme v kódu pracovaly se seznamem startovních políček bez jakýchkoliv dalších detailů. Pro hru ale potřebujeme krom polohy startovních políček ještě několik doplňujících informací, například kterým směrem jsou orientována, a také potřebujeme, aby záznamy zůstávaly ve stejném pořadí. Karolina se rozhodla předělat seznam na slovník, kde klíčem bude číslo políčka a hodnotami budou směr a souřadnice. Starší verze Pythonu ale náhodně mění pořadí položek ve slovníku, bylo tedy třeba použít knihovnu `Collections` a z ní třídu [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict), která starším verzím Pythonu přidává schopnost držet záznamy ve slovníku v takovém pořadí, v jakém jsou do něj vkládány. Od Pythonu 3.6 výše je tato vlastnost slovníků přímo zaintegrovaná do Pythonu. 
```

Reply:
@befeleme V dokumentaci psali Python 3.7 https://docs.python.org/3/library/collections.html#collections.OrderedDict

Reply:
Ok, podle dokumentace:
> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.
Řídila jsem se tím, že nám to předtím fungovalo i v 3.6. :) Souhlasím i s původní 3.7.